tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h

```diff
@@ -0,0 +1,185 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_COMPILER_MLIR_XLA_TRANSFORMS_BUFFER_ASSIGNMENT_H_
+#define TENSORFLOW_COMPILER_MLIR_XLA_TRANSFORMS_BUFFER_ASSIGNMENT_H_
+
+#include "mlir/Analysis/Dominance.h"
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/IR/Builders.h"   // TF:llvm-project
+#include "mlir/IR/Operation.h"  // TF:llvm-project
+#include "mlir/Support/LLVM.h"
+
+namespace mlir {
+namespace xla {
+
+namespace detail {
+/// A specialized dominator analysis that provided access to some private
+/// methods.
+/// TODO(dfki): merge this functionality into the underlying MLIR core dominator
+/// analysis.
+template <bool IsPostDom>
+class BufferAssignmentDominators
+    : public mlir::detail::DominanceInfoBase<IsPostDom> {
+ public:
+  using super = mlir::detail::DominanceInfoBase<IsPostDom>;
+  using super::super;
+
+  /// Finds the nearest common dominator block for the two given blocks first
+  /// and second. If there is no common dominator this function will return
+  /// nullptr.
+  Block* findNearestCommonDominator(Block* first, Block* second) const {
+    assert(first->getParent() == second->getParent() & "Invalid region");
+    return super::dominanceInfos.find(first->getParent())
+        ->second->findNearestCommonDominator(first, second);
+  }
+};
+
+/// A straight-forward alias analysis which ensures that all aliases of all
+/// values will be determined. This is a requirement for the BufferAssignment
+/// class since you need to determine safe positions to place alloc and
+/// deallocs.
+class BufferAssignmentAliasAnalysis {
+ public:
+  using ValueSetT = SmallPtrSet<Value, 16>;
```

Comment:
Will buffer_assignment.* live in tensorflow repo or will it be moved to llvm/mlir? Asking because of the style. It looks pretty prepared to be moved to llvm codebase.

Reply:
Yes and no :) Currently, it would be better to continue the development of the BufferAssignment class in tensorflow since it is particularly designed for XLA/LHLO.